sonarqube/resource_sonarqube_new_code_periods.go

```diff
@@ -0,0 +1,216 @@
+package sonarqube
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+// NewCodePeriods for unmarshalling response body from new_code_periods list definitions.
+type NewCodePeriod struct {
+	Project        string `json:"projectKey"`
+	Branch         string `json:"branchKey"`
+	Type           string `json:"type"`
+	Value          string `json:"value,omitempty"`
+	EffectiveValue string `json:"effectiveValue"`
+	Inherited      bool   `json:"inherited"`
+}
+
+// New Code Period types
+type NewCodePeriodType string
+
+const (
+	SpecificAnalysis NewCodePeriodType = "SPECIFIC_ANALYSIS"
+	PreviousVersion  NewCodePeriodType = "PREVIOUS_VERSION"
+	NumberOfDays     NewCodePeriodType = "NUMBER_OF_DAYS"
+	ReferenceBranch  NewCodePeriodType = "REFERENCE_BRANCH"
+)
+
+// Returns the resource represented by this file.
+func resourceSonarqubeNewCodePeriodsBinding() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceSonarqubeNewCodePeriodsCreate,
+		Read:   resourceSonarqubeNewCodePeriodsRead,
+		Update: resourceSonarqubeNewCodePeriodsCreate,
+		Delete: resourceSonarqubeNewCodePeriodsDelete,
+
+		// Define the fields of this schema.
+		Schema: map[string]*schema.Schema{
+			"branch": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"project": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"type": {
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         false,
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice([]string{string(SpecificAnalysis), string(PreviousVersion), string(NumberOfDays), string(ReferenceBranch)}, false)),
+			},
+			"value": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: false,
+			},
+		},
+	}
+}
+
+func resourceSonarqubeNewCodePeriodsCreate(d *schema.ResourceData, m interface{}) error {
+	sonarQubeURL := m.(*ProviderConfiguration).sonarQubeURL
+	sonarQubeURL.Path = strings.TrimSuffix(sonarQubeURL.Path, "/") + "/api/new_code_periods/set"
+
+	periodType := NewCodePeriodType(d.Get("type").(string))
+	rawQuery := url.Values{
+		"type": []string{string(periodType)},
+	}
+
+	id := "newCodePeriod"
+
+	branch := d.Get("branch").(string)
+	project := d.Get("project").(string)
+	value := d.Get("value").(string)
+
+	// If branch is set, project must also be set
```

Comment:
instead of checking `If branch is set, project must also be set` here we could add this logic to the schema map above. 

Using `RequiredWith` sounds right to me: https://github.com/hashicorp/terraform-plugin-sdk/blob/v2.25.0/helper/schema/schema.go#L312

Reply:
Ahh I wasn't aware of `RequiredWith`, I can definitely make that change